### PR TITLE
feat: Support for headless services

### PR DESCRIPTION
**Pull request by lemaitre-aneo**
Headless services in Kubernetes are services without IPs (neither ClusterIP, nor LoadBalancer). Instead, they require the use of DNS service domain that lists the IPs of all the pods behind the service.
This way, the nginx ingress is aware of the actual pods that are part of the service, and the responsibility to load balance and check health of connections is given back to nginx instead of the L4 load balancer that is a ClusterIP.

This should help with submitter failures as nginx will be able to see from where the failure comes from, and mark this single pod as unhealthy server instead of the whole service. It might also help with single client load balancing and submitter down scaling.

This PR is in conflict with #57.

### CODE REVIEW

File: utils/service-ip/variables.tf [hunk 1]
```diff
@@ -0,0 +1,10 @@
+variable "domain" {
+  description = "Internal domain name of the Kubernetes cluster"
+  type        = string
+  default     = "cluster.local"
+}
+
+variable "service" {
+  description = "Service object"
+  type        = any
+}
```
Comment (lzianekhodja) [inline utils/service-ip/variables.tf:10]: Is it required variable or could we put default = null ?

File: utils/service-ip/variables.tf [hunk 2]
```diff
@@ -0,0 +1,10 @@
+variable "domain" {
+  description = "Internal domain name of the Kubernetes cluster"
+  type        = string
+  default     = "cluster.local"
+}
+
+variable "service" {
```
Comment (lzianekhodja) [inline utils/service-ip/variables.tf:7]: Is it required variable or could we put default = null ?

Comment (lemaitre-aneo) [reply utils/service-ip/variables.tf:7]: Yes, it is required. It is the service object we want to get the IP.

File: utils/service-ip/locals.tf [hunk 3]
```diff
@@ -0,0 +1,4 @@
+locals {
```
Comment (lzianekhodja) [inline utils/service-ip/locals.tf:1]: We can put locals in main.tf.
Pre-commit requires in a module these files: main.tf, variables.tf, outputs.tf, versions.tf and README.md (only a description is recommended in README.md. The documentation about Terraform code are generated).
It is recommaned also give an exemple how to use this module.

File: utils/service-ip/locals.tf [hunk 4]
```diff
@@ -0,0 +1,4 @@
...
+  domain = "${var.service.metadata[0].name}.${var.service.metadata[0].namespace}"
```
Comment (lzianekhodja) [inline utils/service-ip/locals.tf:2]:
```suggestion
  service_domain = "${var.service.metadata[0].name}.${var.service.metadata[0].namespace}"
```

File: utils/service-ip/outputs.tf [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    coalesce(var.service.spec[0].cluster_ip != "None" ? var.service.spec[0].cluster_ip : null),
+
+    # Headless
+    local.fqdn
+  ) : null
+}
+
+output "domain" {
+  description = "Domain Name of the service"
+  value       = local.domain
```
Comment (lzianekhodja) [inline utils/service-ip/outputs.tf:18]:
```suggestion
  value       = local.service_domain
```

File: armonik/external-data.tf [hunk 6]
```diff
@@ -1,74 +1,50 @@
...
-    app_port = kubernetes_service.admin_old_gui[0].spec[0].port[1].port
-    } : {
-    ip       = ""
-    api_port = ""
-    app_port = ""
-  }) : null
+data "kubernetes_config_map" "dns" {
+  metadata {
+    name      = "coredns"
+    namespace = replace(data.kubernetes_secret.deployed_object_storage.id, "/.*/", "kube-system")
```
Comment (lzianekhodja) [inline armonik/external-data.tf:4]: Why secret of deployed_object_storage ?

File: utils/service-ip/locals.tf (hunk 4 above)
Comment (lemaitre-aneo) [reply utils/service-ip/locals.tf]: As the module is for services, it is pretty clear that `domain` refers to the service domain. I changed the input `domain` to `cluster_domain` instead.

File: utils/service-ip/example/main.tf [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    namespace = "kube-system"
+  }
+}
+
+locals {
+  cluster_domain = try(regex("kubernetes\\s+(\\S+)\\s", data.kubernetes_config_map.dns.data["Corefile"])[0], null)
+}
+
+module "my_svc_ip" {
+  source         = "./.."
```
Comment (lzianekhodja) [inline utils/service-ip/example/main.tf:34]:
```suggestion
  source         = "../../service-ip"
```

**Review by lzianekhodja — APPROVED** [review]

**Pull request merged**